server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsInProgress.java

```diff
@@ -76,13 +92,52 @@ public SnapshotDeletionsInProgress withAddedEntry(Entry entry) {
     }
 
     /**
-     * Returns a new instance of {@link SnapshotDeletionsInProgress} which removes
-     * the given entry from the invoking instance.
+     * Returns a new instance of {@link SnapshotDeletionsInProgress} that has the entry with the given {@code deleteUUID} removed from its
+     * entries.
      */
-    public SnapshotDeletionsInProgress withRemovedEntry(Entry entry) {
-        List<Entry> entries = new ArrayList<>(getEntries());
-        entries.remove(entry);
-        return new SnapshotDeletionsInProgress(entries);
+    public SnapshotDeletionsInProgress withRemovedEntry(String deleteUUID) {
+        List<Entry> updatedEntries = new ArrayList<>(entries.size() - 1);
+        boolean removed = false;
+        for (Entry entry : entries) {
+            if (entry.uuid().equals(deleteUUID)) {
+                removed = true;
+            } else {
+                updatedEntries.add(entry);
+            }
+        }
+        return removed ? new SnapshotDeletionsInProgress(updatedEntries) : this;
+    }
+
+    public SnapshotDeletionsInProgress withRemovedSnapshotIds(String repository, Collection<SnapshotId> snapshotIds) {
+        boolean changed = false;
+        List<Entry> updatedEntries = new ArrayList<>(entries.size());
+        for (Entry entry : entries) {
+            if (entry.repository().equals(repository)) {
+                final List<SnapshotId> updatedSnapshotIds = new ArrayList<>(entry.getSnapshots());
+                if (updatedSnapshotIds.removeAll(snapshotIds)) {
+                    changed = true;
+                    updatedEntries.add(entry.withSnapshots(updatedSnapshotIds));
+                } else {
+                    updatedEntries.add(entry);
+                }
+            } else {
+                updatedEntries.add(entry);
+            }
+        }
+        return changed ? new SnapshotDeletionsInProgress(updatedEntries) : this;
+    }
+
+    public SnapshotDeletionsInProgress withRemovedRepository(String repository) {
```

Comment:
If we fail loading repository data for a repository but have queued up deletes we just drop all the deletes for the repo on failure.

Reply:
No if we do this in a CS update we collect the relevant entries in the CS update and fail all the listeners for which we removed the job from the state.
See here https://github.com/elastic/elasticsearch/pull/56911/files#diff-a0853be4492c052f24917b5c1464003dR1512